xml_converter/src/xml_converter.cpp

```diff
@@ -61,30 +63,115 @@ void write_xml_file(string xml_filepath, map<string, Category>* marker_categorie
+        all_categories.add_category()->CopyFrom(proto_category);
+        create_proto_categories(proto_category, &proto_categories);
     }
 
     for (const auto& parsed_poi : *parsed_pois) {
         if (parsed_poi->classname() == "POI") {
             Icon* icon = dynamic_cast<Icon*>(parsed_poi);
             waypoint::Icon poi = icon->as_protobuf();
-            proto_message.add_icon()->CopyFrom(poi);
+            proto_pois.add_icon()->CopyFrom(poi);
+            map_ids.insert(icon->map_id);
         }
         if (parsed_poi->classname() == "Trail") {
             Trail* trail = dynamic_cast<Trail*>(parsed_poi);
             waypoint::Trail poi = trail->as_protobuf();
-            proto_message.add_trail()->CopyFrom(poi);
+            proto_pois.add_trail()->CopyFrom(poi);
+            map_ids.insert(trail->map_id);
         }
     }
-    proto_message.SerializeToOstream(&outfile);
-    outfile.close();
+
+    for (int map_id : map_ids) {
+        ofstream outfile;
+        string output_filepath = proto_filepath + "/" + to_string(map_id) + ".data";
+        outfile.open(output_filepath, ios::out | ios::binary);
+        output_message.CopyFrom(all_categories);
+
+        for (int i = 0; i < proto_pois.icon_size(); i++) {
+            vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
+            if (proto_pois.icon(i).map_id() == map_id) {
+                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
+                if (category == proto_categories.end()) {
+                    cout << "Unknown MarkerCategory " << proto_pois.icon(i).category().name() << endl;
+                }
+                else {
+                    for (string split_category : split_categories) {
+                        category_includes.insert(split_category);
+                    }
+                }
+                output_message.add_icon()->CopyFrom(proto_pois.icon(i));
+            }
+        }
+        for (int i = 0; i < proto_pois.trail_size(); i++) {
+            if (proto_pois.trail(i).map_id() == map_id) {
+                vector<string> split_categories = split(proto_pois.trail(i).category().name(), ".");
+                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
+                if (category == proto_categories.end()) {
+                    cout << "Unknown MarkerCategory " << proto_pois.trail(i).category().name() << endl;
+                }
+                else {
+                    for (string split_category : split_categories) {
+                        category_includes.insert(split_category);
+                    }
+                }
+                output_message.add_trail()->CopyFrom(proto_pois.trail(i));
+            }
+        }
+        for (int i = 0; i < output_message.category_size(); i++) {
+            remove_proto_child(output_message.mutable_category(i), category_includes);
```

Comment:
We should add some explanation why we doing the work to copy all the markers and then removing the ones we don't care about, instead of just only inserting the ones we do care about. Keep in mind doing this means that we have to spent the time to copy all the data, then each time we remove an element from an array we have to copy all the elements that are stored in memory after it to compensate for the now missing element.

Reply:
The comment you added should really have a "we are doing a removal instead of an insertion because ...". This needs to be there because of the redundancy of "add everything then remove it" instead of the much simpler "only add what we need". The extra cost needs to be clearly  justifiable.

Reply:
How does this sound instead 

```        
        // Marker_Categories has a tree hierarchy while POIS have a flat hierarchy.
        // We are doing a removal instead of an insertion because each parent category contains the data for all of its children.
        // It would be impractical to include all of the data from each category except for the children and then re-add the children. 
        // That would require coping every non-child attribute individually and iterating over all the children.
        // This pruning method is slower but ensures that the all wanted information is kept.
```

Reply:
Definitely better and makes clear the issue of not being able to "copy everything but".

However now it raises a new question which is: Why is the data formatted like that to begin with if its intention is to be serialized here?

Reply:
Keep in mind this is not me telling you that the entire structure of the code needs to be uprooted, it just needs to be explained here, even if the explanation is "we should uproot the entire structure of the code, but we are not going to do that right now". All that is needed is a logical thought through blurb so the next developer who reads it can understand why it is the way it is. For example it might be "we need to copy all the data be cause protobuf has a memory corruption issue if we try to share even a single string", this is not the case of course but a fresh pair of eyes wont know that right away and might just assume the code should up refactored only to run into said string corruption issue down the line. But if there is a comment saying "the code should be restructured in the future but we are doing this now because it should still be 'fast enough'" then the future developer knows that the reason it is being done in a weird way can be resolved with said refactor.